### PR TITLE
[processor/tailsampling] Fix SpanCount sampler not using the correct span count for decision

### LINKED ISSUE #13865: [processor/tailsampling] SpanCount sampler makes incorrect decision on high volume of spans
**Describe the bug**
Sometimes, the SpanCount sampler makes the incorrect decision even though there's many spans coming in for the same trace id. This is more likely to happen when there's a high volume of spans and traces coming in to the processor.

After looking at the code, it seems this might be caused by the internal batching behaviour of the processor.

**What version did you use?**
v0.54.0

**What config did you use?**
``` yaml
processors:
  tail_sampling:
    decision_wait: 10s
    num_traces: 100
    expected_new_traces_per_sec: 10
    policies:
      [
          {
            name: test-policy-8,
            type: span_count,
            span_count: {min_spans: 2}
          }
      ]
```

**Environment**
Docker image 
  OS: Alpine
Compiler(if manually compiled): go 1.18.3

### PR DESCRIPTION
**Pull request by Symbianx**
Fixes #13865
**Description:**
There is a bug with the SpanCount sampler. If spans are put in different batches, it may make a wrong decision.

e.g.
With SpanCount minSpans set to `2`; if 2 spans of the same trace are put into different batches, the trace will not be sampled.

### CODE REVIEW

**Comment by Symbianx** [comment]
@dmitryax can you approve the the running workflows and take a look at the PR? Thank you!

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler.go [hunk 1]
```diff
@@ -38,14 +38,8 @@ func NewSpanCount(logger *zap.Logger, minSpans int32) PolicyEvaluator {
...
 
-	traceData.Lock()
-	batches := traceData.ReceivedBatches
-	traceData.Unlock()
-
-	for _, trace := range batches {
-		if trace.SpanCount() >= int(c.minSpans) {
-			return Sampled, nil
-		}
+	if int(traceData.SpanCount.Load()) >= int(c.minSpans) {
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/span_count_sampler.go:41]: Could you please add a new test, to demonstrate the problem and ensure we don't have a regression in the future?

**Comment by jpkrohling** [comment]
Is there a GitHub issue tracking this?

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler.go (hunk 1 above)
Comment (Symbianx) [reply processor/tailsamplingprocessor/internal/sampling/span_count_sampler.go:41]: Looks like 1 of the tests was already covering this but with the incorrect expected outcome. Fixed it and should now cover this logic.

**Comment by Symbianx** [comment]
@jpkrohling there wasn't but I created #13865 to track it.

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go [hunk 2]
```diff
@@ -41,23 +42,30 @@ func TestEvaluate_NumberSpans(t *testing.T) {
...
+				3,
 			},
-			NotSampled,
+			Sampled,
 		},
 		{
-			"Same number of spans than the threshold",
+			"Same number of spans as the threshold, in one single batch",
 			[]int32{
-				1, 2, 1,
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go:53]: Please, keep the previous cases.

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go [hunk 3]
```diff
@@ -41,23 +42,30 @@ func TestEvaluate_NumberSpans(t *testing.T) {
...
+			[]int32{
+				1, 1, 1,
 			},
 			Sampled,
 		},
 		{
-			"Bigger amount of spans than the threashold",
+			"Same number of spans as the threshold, across multiple batches",
 			[]int32{
-				1, 3, 1,
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go:60]: Please, reinstate this one as well.

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go (hunk 2 above)
Comment (Symbianx) [reply processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go:53]: done.

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go (hunk 3 above)
Comment (Symbianx) [reply processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go:60]: done.

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go [hunk 4]
```diff
@@ -41,26 +42,47 @@ func TestEvaluate_NumberSpans(t *testing.T) {
...
+				2,
+			},
+			Sampled,
+		},
+		{
+			"Bigger amount of spans than the threshold, across multiple batches",
+			[]int32{
+				1, 1, 1,
+			},
+			Sampled,
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go:70]: Previously, this was not sampled. This was the bug, right?

**Review by jpkrohling — APPROVED** [review]

File: processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go (hunk 4 above)
Comment (Symbianx) [reply processor/tailsamplingprocessor/internal/sampling/span_count_sampler_test.go:70]: Yes, that's right

**Pull request merged**